iModelJsNodeAddon/JsInterop.cpp

```diff
@@ -922,6 +931,80 @@ DbResult JsInterop::ImportSchemas(DgnDbR dgndb, bvector<Utf8String> const& schem
     return dgndb.SaveChanges();
     }
 
+//---------------------------------------------------------------------------------------
+// @bsimethod
+//---------------------------------------------------------------------------------------
+DbResult JsInterop::ConvertEC2XmlSchemas(bvector<Utf8String> const& ec2XmlStrings, bvector<Utf8String>& ec3XmlStrings, ECSchemaReadContextPtr schemaContext)
+    {
+    if (0 == ec2XmlStrings.size())
+        return BE_SQLITE_ERROR;
+
+    if (schemaContext.IsNull())
+        schemaContext = ECSchemaReadContext::CreateContext(false /*=acceptLegacyImperfectLatestCompatibleMatch*/, true /*=includeFilesWithNoVerExt*/);
+
+    JsInterop::AddFallbackSchemaLocaters(schemaContext);
+
+    bvector<bpair<SchemaKey, ECSchemaPtr>> schemaKeyPairs;
+    StringSchemaLocater locater;
+    for (Utf8String ec2XmlString : ec2XmlStrings)
+        {
+        SchemaKey key;
+        SchemaReadStatus status = ECSchema::ReadSchemaKey(ec2XmlString, key);
+        if (SchemaReadStatus::Success != status)
+            return BE_SQLITE_ERROR;
+        locater.AddSchemaString(key.GetName(), ec2XmlString);
+        schemaKeyPairs.push_back(std::make_pair(key, nullptr));
+        }
+    schemaContext->AddSchemaLocater(locater);
+
+    for (int i = 0; i < ec2XmlStrings.size(); i++)
+        {
+        Utf8String ec2XmlString = ec2XmlStrings[i];
+        ECSchemaPtr schema;
+        SchemaReadStatus schemaStatus = ECSchema::ReadFromXmlString(schema, ec2XmlString.c_str(), *schemaContext);
+
+        if (SchemaReadStatus::DuplicateSchema == schemaStatus && schemaKeyPairs[i].second == nullptr)
+        {
+            schema = ECSchema::LocateSchema(schemaKeyPairs[i].first, *schemaContext);

```

Comment:
I don't think this is necessary. The incoming schemas are not in dependency order, so loading one of them may automatically load its referenced schemas into the context, and when you call ReadFromXmlString() again, you will run into this DuplicateSchema problem and use the one from the context instead.

Instead of doing this, you could just always call schemaContext.LocateSchema() and not call ReadFromString at all. It should boil down to the same logic and you would not need this additional block here.

Reply:
Updated the code to use the ECSchema::LocateSchema() method. The schema is located using SchemaMatchType::LatestWriteCompatible as the match type criteria.